File: OOPSolution.py
from enum import Enum
from collections import defaultdict

class Bag(Enum):
    SMALL = (5, 0.25)
    MEDIUM = (10, 0.5)
    LARGE = (15, 1)
    
    def __init__(self, price, size):
        self.price = price
        self.size = size

# # Alternatively something like this 
# class Bag:
#     def __init__(self, name, size, price):
#         self.name = name
#         self.size = size
#         self.price = price 

class Van:
    
    def __init__(self, id, maxCapacity):
        self.id = id
        self.maxCapacity = maxCapacity
        self.currCapacity = 0
        self.pickupCount = 0
    
    def canFit(self, bag):
        return self.currCapacity + bag.size <= self.maxCapacity
    
    def addBag(self, bag):
        self.currCapacity += bag.size
        self.pickupCount += 1
    
    def getGasCost(self):
        return 4 * (0.1 * self.pickupCount * (self.currCapacity + self.maxCapacity))

class Laundry:

    def __init__(self):
        self.revenue = 0
        self.vans = [Van("Eric's Van", 10)]
    
    def singlePickup(self, address, bag):
        # Part 2 
        if self.van.currCapacity + bag.size <= self.van.maxCapacity:
            self.van.currCapacity += bag.size
            self.revenue += bag.price
        else:
            raise Exception("Max capacity reached, sad.")
    
    def pickup(self, address, bag):
        # Part 3
        for van in self.vans:
            if van.canFit(bag):
                van.addBag(bag)
                self.revenue += bag.price
                return
    
    def getNetIncome(self):
        gas_cost = sum(van.getGasCost() for van in self.vans)
        return self.revenue - gas_cost

class LaundryCorp:
    
    def __init__(self):
        self.franchises = defaultdict(Laundry)
    
    def addFranchise(self, name, laundry):
        self.franchises[name] = laundry
    
    def getLargestRevenueFranchise(self):
        return max(self.franchises, key=lambda name: self.franchises[name].revenue, default="")
        

if __name__ == "__main__":
    # Some tests, not necessary for actual interview 
    clean4impact = Laundry()
    clean4impact.pickup("H4I Road", Bag.SMALL)
    assert(clean4impact.revenue == 5)
    assert(clean4impact.vans[0].currCapacity == 0.25)
    assert(round(clean4impact.getNetIncome(), 2) == 0.9)
    cornell = Laundry()
    hack4impact = LaundryCorp()
    hack4impact.addFranchise("UIUC", clean4impact)
    hack4impact.addFranchise("Cornell", cornell)
    assert(hack4impact.getLargestRevenueFranchise() == "UIUC")

File: OOPSolution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

// [!] How do you imagine your code will obtain laundry requests, and from where will they originate?
//      I imagine my code will obtain laundry requests from some frontend via a REST API which then calls my code.
// [!] How do you imagine your code will present computed information (revenue, net income), or how will it send this information to a presentation system?
//      In progress
// [!] How do you plan to deploy your program as a service accessible to other business systems?

// [!] Let's say we put all these things into production, what's the worst that could happen?

public class OOPSolution {
    public static void main(String[] args) {
    }

    public static class LaundryCorporation {
        private Map<String, Store> franchises = new HashMap<>();

        public String getHighestRevenueFranchise() {
            String highestRevenueFranchise = franchises.keySet().iterator().next();
            for (Entry<String, Store> entry : franchises.entrySet()) {
                if (entry.getValue().getRevenue() > franchises.get(highestRevenueFranchise).getRevenue()) {
                    highestRevenueFranchise = entry.getKey();
                }
            }
            return highestRevenueFranchise;
        }
    }

    public static class Store {
        private int revenue = 0;
        private List<LaundryVan> vans = new ArrayList<LaundryVan>();

        public void handlePickup(String address, LaundryBag bag) {
            for (LaundryVan van : vans) {
                if (van.canFit(bag)) {
                    van.addPickup(bag);
                    revenue += bag.getPrice();
                    return;
                }
            }
        }

        public int getRevenue() {
            return revenue;
        }

        public double getNetIncome() {
            double totalGasCost = 0;
            for (LaundryVan van : vans) {
                totalGasCost += van.getGasCost();
            }
            return revenue - totalGasCost;
        }
    }

    public class LaundryVan {
        private int id;
        private int capacity;
        private int capacityUsed = 0;
        private int numberOfPickups = 0;

        public LaundryVan(int id, int capacity) {
            this.id = id;
            this.capacity = capacity;
        }

        public int getId() {
            return id;
        }

        public boolean canFit(LaundryBag bag) {
            return capacityUsed + bag.getSize() <= capacity;
        }

        public void addPickup(LaundryBag bag) {
            capacityUsed += bag.getSize();
            numberOfPickups++;
        }

        public double getGasCost() {
            double gasUsed = 0.1 * numberOfPickups * (this.capacityUsed + this.capacity);
            return 4 * gasUsed;
        }
    }

    public static enum LaundryBag {
        SMALL(5, 0.25),
        MEDIUM(10, 0.5),
        LARGE(15, 1);

        private int price;
        private double size;

        private LaundryBag(int price, double size) {
            this.price = price;
            this.size = size;
        }

        public int getPrice() {
            return price;
        }

        public double getSize() {
            return size;
        }
    }
}
